File: src/challenges/java/minimalheighttree/MinimalHeightTree.java
package challenges.java.minimalheighttree;


import challenges.java.TreeNode;

public class MinimalHeightTree {
    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 4, 5, 6};
        printTreeInOrder(minimalHeight(arr));
    }

    private static TreeNode minimalHeight(int[] arr) {
        return addToTree(arr, 0, arr.length - 1);
    }

    private static TreeNode addToTree(int[] arr, int start, int end) {
        if (start > end) {
            return null;
        }
        int mid = end - start / 2;
        TreeNode middle = new TreeNode(arr[mid]);
        middle.left = addToTree(arr, start, mid - 1);
        middle.right = addToTree(arr, mid + 1, end);
        return middle;
    }

    private static void printTreeInOrder(TreeNode root) {
        if (root != null) {
            if (root.left != null) {
                printTreeInOrder(root.left);
            }
            System.out.println(root.data);
            if (root.right != null) {
                printTreeInOrder(root.right);
            }
        }
    }
}

File: src/trees/kotlin/minimalheighttree/MinimalHeightTree.kt
package trees.kotlin.minimalheighttree

import trees.kotlin.TreeNode

fun main() {
    val arr = arrayOf(1, 2, 3, 4, 5, 6, 7, 8, 9)
    printTreeInOrder(minimalHeightTree(arr))
}

fun minimalHeightTree(arr: Array<Int>): TreeNode? {
    return addToTree(arr, 0, arr.size - 1)
}

fun addToTree(arr: Array<Int>, start: Int, end: Int): TreeNode? {
    if (start > end) return null
    val mid = (start + end) / 2
    val middle = TreeNode(arr[mid])
    middle.left = addToTree(arr, start, mid - 1)
    middle.right = addToTree(arr, mid + 1, end)
    return middle
}

fun printTreeInOrder(root: TreeNode?) {
    root?.let {
        root.left?.let { left ->
            printTreeInOrder(left)
        }
        println(root.data)
        root.right?.let { right ->
            printTreeInOrder(right)
        }
    }
}
